### PR TITLE
BREAKING: Use @metamask/utils

### PR DESCRIPTION
**Pull request by rekmarks**
Replace various utility types and functions with implementations copied over to `@metamask/utils`. Functionality should be practically identical.

The one possible deviation is that two `in` checks have been replaced with `hasProperty`, which does not walk the prototype chain. This should not be breaking for our purposes.

### CODE REVIEW

File: src/JsonRpcEngine.ts [hunk 1]
```diff
@@ -487,7 +441,7 @@ export class JsonRpcEngine extends SafeEventEmitter {
     res: PendingJsonRpcResponse<unknown>,
     isComplete: boolean,
   ): void {
-    if (!('result' in res) && !('error' in res)) {
+    if (!hasProperty(res, 'result') && !hasProperty(res, 'error')) {
```
Comment (rekmarks) [inline src/JsonRpcEngine.ts:444]: The replaced `in` checks.

File: src/JsonRpcEngine.ts [hunk 2]
```diff
@@ -79,9 +29,9 @@ export type JsonRpcEngineEndCallback = (
   error?: JsonRpcEngineCallbackError,
 ) => void;
 
-export type JsonRpcMiddleware<T, U> = (
-  req: JsonRpcRequest<T>,
-  res: PendingJsonRpcResponse<U>,
+export type JsonRpcMiddleware<Params, Result> = (
```
Comment (mcmire) [inline src/JsonRpcEngine.ts:32]: I like these type parameter names a lot better than `T` and `U` 👍🏻

File: src/JsonRpcEngine.ts [hunk 3]
```diff
@@ -1,69 +1,19 @@
...
-  error: JsonRpcError;
-}
-
-export type JsonRpcResponse<T> = JsonRpcSuccess<T> | JsonRpcFailure;
-
-export interface PendingJsonRpcResponse<T> extends JsonRpcResponseBase {
-  result?: T;
-  error?: Error | JsonRpcError;
-}
+export type PendingJsonRpcResponse<Result> = Omit<
```
Comment (mcmire) [inline src/JsonRpcEngine.ts:10]: Not suggesting any changes, but I wonder if this is type is right, considering that it seems that such a response can theoretically have both `result` and `error`.

**Review by mcmire — APPROVED** [review]
Seems good to me!

File: src/JsonRpcEngine.ts (hunk 3 above)
Comment (rekmarks) [reply src/JsonRpcEngine.ts:10]: The type is meant to be used inside middleware functions and internally in `JsonRpcEngine` during request processing. It permits both, either, or neither property being present. Or did you mean something else?

**Pull request merged**